packages/cli/src/parallel-tests/parallel-tests.handler.ts

```diff
@@ -102,75 +72,15 @@ export const runAllTestsInParallel: (
   const maxTasks = parallelTasks ?? Math.max(cpus().length, 1);
   logger.debug(`Running with ${maxTasks} maximum tasks in parallel`);
 
-  const taskHandler = join(__dirname, "task.handler.js");
-
   // Starts running a test case in a child process
   const startTask = (rerunnableTestCase: RerunnableTestCase) => {
     const { id, ...testCase } = rerunnableTestCase;
-    const deferredResult = defer<DetailedTestCaseResult>();
-    const child = fork(taskHandler, [], { stdio: "inherit" });
-
-    const messageHandler = (message: unknown) => {
-      if (
-        message &&
-        typeof message === "object" &&
-        (message as any)["kind"] === "result"
-      ) {
-        const resultMessage = message as ResultMessage;
-        deferredResult.resolve(resultMessage.data.result);
-        child.off("message", messageHandler);
-      }
-    };
-
-    child.on("error", (error) => {
-      if (deferredResult.getState() === "pending") {
-        deferredResult.reject(error);
-      }
-    });
-    child.on("exit", (code) => {
-      if (code) {
-        logger.debug(`child exited with code: ${code}`);
-      }
-      if (deferredResult.getState() === "pending") {
-        deferredResult.reject(new Error("No result"));
-      }
-    });
-    child.on("message", messageHandler);
-
-    // Send test case and arguments to child process
     const isRetry = resultsByTestId.has(id);
-    const initMessage: InitMessage = {
-      kind: "init",
-      data: {
-        logLevel: logger.getLevel(),
-        dataDir: getMeticulousLocalDataDir(),
-        replayOptions: {
-          apiToken,
-          commitSha,
-          testCase,
-          deflake,
-          replayTarget: getReplayTargetForTestCase({
-            useAssetsSnapshottedInBaseSimulation,
-            appUrl,
-            testCase,
-          }),
-          executionOptions,
-          screenshottingOptions,
-          generatedBy: { type: "testRun", runId: testRun.id },
-          testRunId: testRun.id,
-          replayEventsDependencies,
-          suppressScreenshotDiffLogging: isRetry,
-        },
-      },
-    };
-    child.send(initMessage);
 
     // Handle task completion
-    deferredResult.promise
+    executeTest(testCase, isRetry)
       .catch(() => null)
       .then(async (result) => {
-        --inProgress;
```

Comment:
Bug #1: We were decrementing this too early, which meant a 2nd promise could call allTasksDone.resolve() while we're still saving results for the 1st promise